### PR TITLE
J3DMatBlock

### PR DESCRIPTION
**Pull request by Jcw87**
I was working on this when I noticed #83.

### CODE REVIEW

File: include/dolphin/gd/GDBase.h [hunk 1]
```diff
@@ -40,6 +40,10 @@ inline u8* GDGetCurrPointer() {
     return __GDCurrentDL->ptr;
 }
 
+inline s32 GDGetCurrOffset() {
```
Comment (encounter) [inline include/dolphin/gd/GDBase.h:43]: Is this real? I thought I only saw GDGetGDLObjOffset in the debug map

File: src/JSystem/J3DGraphBase/J3DGD.cpp [hunk 2]
```diff
@@ -7,121 +7,129 @@
...
+void J3DGDSetTevOrder(GXTevStageID stage, GXTexCoordID coord1, GXTexMapID map1, GXChannelID color1, GXTexCoordID coord2, GXTexMapID map2, GXChannelID color2) {
     /* Nonmatching */
+    static u8 c2r[] = {0, 1, 0, 1, 0, 1, 7, 5, 6, 0, 0, 0, 0, 0, 0, 7};
+
+    GXTexCoordID r31 = coord1 >= 8 ? GX_TEXCOORD0 : coord1;
+    GXTexCoordID r30 = coord2 >= 8 ? GX_TEXCOORD0 : coord2;
+    GDOverflowCheck(5);
+    bool r27 = map1 != GX_TEXMAP_NULL && (map1 & GX_TEXMAP_DISABLE) == 0;
+    bool r26 = map2 != GX_TEXMAP_NULL && (map2 & GX_TEXMAP_DISABLE) == 0;
+    J3DGDWriteBPCmd((stage / 2 + 0x28) << 24 | r31 << 3 | r27 << 6 | c2r[color2 & 0xf] << 19 | r26 << 18 | r30 << 15 | (map2 & 7) << 0xc | c2r[color1 & 0xf] << 7 | r27 << 6 | r31 << 3 | map1 & 7);
```
Comment (encounter) [inline src/JSystem/J3DGraphBase/J3DGD.cpp:114]: Good lord

File: src/JSystem/J3DGraphBase/J3DMatBlock.cpp [hunk 3]
```diff
@@ -30,6 +32,30 @@ static int SizeOfJ3DColorBlockLightOffLoad = SizeOfLoadMatColors + SizeOfLoadCol
 static int SizeOfJ3DColorBlockAmbientOnLoad =
     SizeOfLoadMatColors + SizeOfLoadAmbColors + SizeOfLoadColorChans;
 
+inline void loadMatColors(const J3DGXColor* color) {
```
Comment (encounter) [inline src/JSystem/J3DGraphBase/J3DMatBlock.cpp:35]: These make sense, I must have missed them in the debug map

**Review by encounter — APPROVED** [review]
Nice! You got a lot further than I did

File: include/dolphin/gd/GDBase.h (hunk 1 above)
Comment (Jcw87) [reply include/dolphin/gd/GDBase.h:43]: I'm working off of the twilight princess debug rom. Some areas are very similar, J3DMatBlock included. GDGetCurrOffset does exist in the wind waker debug map, so I see no reason not to use it.

File: src/JSystem/J3DGraphBase/J3DGD.cpp (hunk 2 above)
Comment (Jcw87) [reply src/JSystem/J3DGraphBase/J3DGD.cpp:114]: That part doesn't even match yet. The massive `or` chains are difficult to parse, even on the twilight princess debug rom.

Comment (magcius) [reply src/JSystem/J3DGraphBase/J3DGD.cpp:114]: You can probably piece it together from https://github.com/devkitPro/libogc/blob/master/libogc/gx.c#L3965-L4008

Comment (Jcw87) [reply src/JSystem/J3DGraphBase/J3DGD.cpp:114]: It's the ordering of the terms that gets me. I try to re-order them, and then the compiler will just suddenly change the order of everything. I don't have the patience to trial and error on that.

**Pull request merged**